src/storage-provider.js

```diff
@@ -0,0 +1,95 @@
+'use strict';
+let debugModule = require('debug');
+let assert = require('assert');
+
+/**
+ * Abstract base class for StorageProviders.  Every method of this abstract
+ * base class may be overridden for each provider aside from the constructor.
+ * The constructor should only take a single object as the configuration so
+ * that we can use typed-env-config as the sole system for configuring them.
+ * It is permissible and recommended that implementors validate the config
+ * object.  This class should only deal with storage provider operations and
+ * not any sort of caching or input functions, aside from extra input
+ * validation required by the backend
+ */
+class StorageProvider {
+
+  /**
+   * Store the configuration for this storage provider in the variable
+   * 'this.config' for use in following methods.
+   */
+  constructor(config) {
+    assert(config, 'must pass a configuration object to StorageProvider constructor');
+    assert(config.id, 'all StorageProviders must have an ID');
+    this.id = config.id;
+    this.debug = debugModule(`cloud-mirror:${this.constructor.name}:${this.id}`);
+  }
+
+  /**
+   * Perform any async init code require for this storage provider.  Default action
+   * is for there to be no init code.
+   */
+  async init() { }
+
+  /**
+   * Store a file with the StorageProvider.  This must be overridden with a method
+   * which takes:
+   *   - rawUrl: the original input url as plain text
+   *   - inputStream: a ReadableStream for the object to store
+   *   - headers: HTTP headers required to be set on the object in the storage
+   *     provider
+   *   - storageMetadata: metadata that should be stored in the storage provider
+   *
+   * and inserts the object into the storage provider
+   */
+  async put(rawUrl, inputStream, headers, storageMetadata) {
+    throw new Error('This StorageProvider implementation must implement .put()');
+  }
+
+  /**
+   * Remove an internal address from the storage provider
+   */
+  async purge(internalAddress) {
+    throw new Error('This StorageProvider implementation must implement .purge()');
+  }
+
+  /**
+   * Perform any extra input validation required by the backend
+   */
+  async validateInput(rawUrl) {
+    return true;
+  }
+
+  /**
+   * Take the request-promise response from an HTTP HEAD request for an object
+   * in this storage provider and return a Date object which represents when
+   * this object will be cleaned up from the cache
+   */
+  async expirationDate(response) {
+    throw new Error('This StorageProvider implementation must implement .expirationDate()');
+  }
+
+  /**
+   * A world address is what we will eventually redirect to.  This method
+   * should map an internal address to a world address
+   */
+  worldAddress(internalAddress) {
```

Comment:
well, either the function is called interalAddress(rawUrl) or it's called encodeUrl(rawUrl)...  In the single place where we actually call worldAddress, the difference here boils down to:

```javascript
worldAddress(internalAddress(rawUrl))
```
or
```javascript
worldAddress(encodeURIComponent(rawUrl))
```
I really don't see what the problem here is.  For most cases, we will use the default internalAddress() implementation.  Having this layer of indirection let's do interesting things with naming objects.  Say for Azure, which does not expire artifacts, we wanted to make the object name be:

```javascript
new Date().toStringSortable() + encodeURIComponent(rawUrl) // ignore made up date method
```
would could do that here with the help of a persistence tool.  Then the expiry program could list all objects which start with, e.g., 2016-03-01 and delete like that.  

Say we wanted to have a private artifact store and wanted to obfuscate the nameing, we could do that using a persistence layer in the storageprovider's internal address.

Furthermore, removing this function wouldn't be removing it substantively, it would merely change how we call worldAddress and would be duplicating the logic of internalAddress inside of worldAddress.  Right now we have one single method per implementation that tells us how to take a rawUrl and give us the key for the object storage.  If we do what you're suggesting, we'll have two methods which tell us how to take a rawUrl and give us the key for the object storage.  There are no benefits from removing this function and only downsides.  Mapping rawUrl => object store name and mapping object  store name => public URL are two distinct operations which I don't feel we should be confusing.